smolyay/surrogate.py

```diff
@@ -0,0 +1,272 @@
+
+        Parameters
+        ----------
+        real_function: callabe
+            Function to be approximated.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition.
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+        """
+        # transform grid points
+        transformed_grids = numpy.zeros((
+            numpy.array(self.grid.points).shape[0], self._domain.shape[0]))
+        for dimension in range(self._domain.shape[0]):
+            transformed_grids[:, dimension] = (
+                numpy.polynomial.polyutils.mapdomain(
+                    numpy.array(self.grid.points)[:, dimension],
+                    (-1, 1), self.domain[dimension]))
+
+        # evaluate real function at grid points
+        real_function_at_grids = numpy.zeros((numpy.array(
+            self.grid.points).shape[0]))
+        for point_num in range(numpy.array(self.grid.points).shape[0]):
+            real_function_at_grids[point_num] = (
+                real_function(*transformed_grids[point_num]))
+
+        # fit
+        return self.train_from_data(real_function_at_grids, linear_solver)
+
+    def train_from_data(self, real_function_at_grids, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to data.
+
+        Parameters
+        ----------
+        real_function_at_grids: list
+            Real function at grid points.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+
+        Raises
+        ------
+        IndexError
+            Data should be of length of the grid points.
+        """
+        real_function_at_grids = numpy.array(real_function_at_grids, ndmin=2)
+        if real_function_at_grids.shape[1] != (
+                numpy.array(self.grid.points).shape[0]):
+            raise IndexError("Data must be in length of the grid points.")
+
+        if linear_solver == 'lu':
+            self._coefficients = numpy.linalg.solve(
+                self._make_basis_matrix(),
+                real_function_at_grids).tolist()
+        elif linear_solver == 'inv':
+            self._coefficients = numpy.dot(numpy.linalg.inv(
+                self._make_basis_matrix()),
+                real_function_at_grids).tolist()
+        elif linear_solver == 'lstsq':
+            self._coefficients = numpy.linalg.lstsq(self._make_basis_matrix(),
+                                                    real_function_at_grids,
+                                                    rcond=None)[0].tolist()
```

Comment:
Make sure to catch unknown solvers, and reduce duplication:
```suggestion
        basis_matrix = self._make_basis_matrix()
        if linear_solver == 'lu':
            self._coefficients = numpy.linalg.solve(basis_matrix, data)
        elif linear_solver == 'inv':
            self._coefficients = numpy.dot(numpy.linalg.inv(basis_matrix), data)
        elif linear_solver == 'lstsq':
            self._coefficients = numpy.linalg.lstsq(basis_matrix, data, rcond=None)[0]
        else:
            raise ValueError('Solver not recognized')
```
You can call the `tolist()` method in the coefficients getter if you so desire, no need to use it here since it will probably be convenient to keep this as a numpy array.